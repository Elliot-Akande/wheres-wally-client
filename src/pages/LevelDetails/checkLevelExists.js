const checkLevelExists = (level) => {
  return 0 < level && level <= 4;
};

export default checkLevelExists;
