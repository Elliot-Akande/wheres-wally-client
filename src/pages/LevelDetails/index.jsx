import { Link, useParams } from "react-router-dom";
import Leaderboard from "../../components/Leaderboard/Leaderboard";
import useFetch from "../../hooks/useFetch";

const LevelDetails = () => {
  const { levelNum } = useParams();
  const { data, loading, error } = useFetch(`/levels/${levelNum}/leaderboard`);

  if (!error) {
    return (
      <>
        <h1>level {levelNum}</h1>
        <Link to={`/level/${levelNum}`}>Start</Link>
        <Leaderboard data={data} loading={loading} />
        <Link to={`/`}>Back</Link>
      </>
    );
  }

  if (error.status === 404) {
    return (
      <div>
        <h2>This level doesn&apos;t exist!</h2>
        <Link to="/">Back to home page</Link>
      </div>
    );
  }

  return (
    <div>
      <h2>There was an error fetching level data!</h2>
      <Link to="/">Back to home page</Link>
    </div>
  );
};

export default LevelDetails;
