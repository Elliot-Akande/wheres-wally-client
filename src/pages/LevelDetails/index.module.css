.container {
  margin: 48px 0;
  padding: 32px;

  display: flex;
  flex-direction: column;
  gap: 16px;

  background-color: var(--neutral-700);
  border-radius: 16px;
}

.header {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.title {
  margin: 0;
}

.start,
.back {
  padding: 8px 16px;

  font-weight: bold;
  text-decoration: none;
  border-radius: 30px;

  transition: 0.1s;
}

.start {
  color: var(--neutral-100);
  background: linear-gradient(180deg, var(--accent), var(--accent-dark));
}

.back {
  align-self: center;
  color: var(--neutral-100);
  background-color: var(--neutral-500);
}

.start:hover,
.back:hover {
  transform: scale(1.05);
}
