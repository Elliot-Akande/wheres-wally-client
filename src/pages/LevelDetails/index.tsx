import { Link, useParams } from "react-router-dom";
import Leaderboard from "../../components/Leaderboard/Leaderboard";
import useFetch from "../../hooks/useFetch";
import {
  BackButton,
  Heading,
  StartButton,
  StyledLevelDetails,
  Top,
} from "./styles";

interface Data {
  scores: Array<{
    _id: string;
    name: string;
    score: number;
  }>;
}

const LevelDetails = () => {
  const { levelNum } = useParams();
  const { data, loading, error } = useFetch<Data>(
    `/levels/${levelNum}/leaderboard`
  );

  if (error) {
    if (error.message === "Not found.") {
      return (
        <div>
          <h1>Error</h1>
          <p>This level doesn&apos;t exist!</p>
          <Link to="/">Back to home page</Link>
        </div>
      );
    }

    return (
      <div>
        <h1>Error</h1>
        <p>There was a problem fetching level data!</p>
        <Link to="/">Back to home page</Link>
      </div>
    );
  }

  return (
    <StyledLevelDetails>
      <Top>
        <Heading>LEVEL {levelNum}</Heading>
        <StartButton to={`/level/${levelNum}`}>Start</StartButton>
      </Top>
      <Leaderboard data={data} loading={loading} />
      <BackButton to={`/`}>Back</BackButton>
    </StyledLevelDetails>
  );
};

export default LevelDetails;
