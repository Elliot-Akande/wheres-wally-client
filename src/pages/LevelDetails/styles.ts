import { Link } from "react-router-dom";
import { css, styled } from "styled-components";

export const StyledLevelDetails = styled.div`
  margin: 48px 0;
  padding: 32px;

  display: flex;
  flex-direction: column;
  gap: 16px;

  background-color: var(--neutral-700);
  border-radius: 16px;
`;

export const Top = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

export const Heading = styled.h1`
  margin: 0;
`;

const buttonStyles = css`
  padding: 8px 16px;

  font-weight: bold;
  text-decoration: none;
  border-radius: 30px;

  transition: 0.1s;

  &:hover {
    transform: scale(1.05);
  }
`;

export const StartButton = styled(Link)`
  ${buttonStyles}

  color: var(--neutral-100);
  background: linear-gradient(180deg, var(--accent), var(--accent-dark));
`;

export const BackButton = styled(Link)`
  ${buttonStyles}

  align-self: center;
  color: var(--neutral-100);
  background-color: var(--neutral-500);
`;
