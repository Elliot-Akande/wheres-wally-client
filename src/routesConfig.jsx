import App from "./App.jsx";
import Error from "./pages/Error/index.jsx";
import Home from "./pages/Home/index.jsx";
import LevelDetails from "./pages/LevelDetails/index.jsx";
import Game from "./pages/Game/index/index.jsx";

const routesConfig = [
  {
    path: "/",
    element: <App />,
    errorElement: <Error />,
    children: [
      { index: true, element: <Home /> },
      { path: "level/:levelId/details", element: <LevelDetails /> },
      { path: "level/:levelId", element: <Game /> },
    ],
  },
];

export default routesConfig;
