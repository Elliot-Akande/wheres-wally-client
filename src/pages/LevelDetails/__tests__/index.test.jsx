import { render } from "@testing-library/react";
import {
  MemoryRouter,
  RouterProvider,
  createMemoryRouter,
  useParams,
} from "react-router-dom";
import { describe, expect, it, vi } from "vitest";
import LevelDetails from "../index.jsx";
import routesConfig from "/src/routesConfig.jsx";

const mocks = vi.hoisted(() => {
  return {
    useParams: vi.fn(),
    checkLevelExists: vi.fn(),
  };
});

vi.mock("../checkLevelExists.js", () => ({
  default: mocks.checkLevelExists,
}));

vi.mock("react-router-dom", async () => {
  const mod = await vi.importActual("react-router-dom");
  return {
    ...mod,
    useParams: mocks.useParams,
  };
});

vi.mock("/src/components/Leaderboard/Leaderboard.jsx", () => ({
  default: () => <div>Leaderboard Component</div>,
}));

describe("LevelDetails", () => {
  it("renders title, start link, leaderboard, and home link.", () => {
    const levelNum = 3;
    mocks.useParams.mockReturnValue({ levelNum });

    const { container } = render(
      <MemoryRouter>
        <LevelDetails />
      </MemoryRouter>
    );

    expect(useParams).toBe(mocks.useParams);
    expect(useParams).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });

  it("redirects to home page on levelNum that doesn't exist.", () => {
    const levelNum = 1;
    mocks.useParams.mockReturnValue({ levelNum });
    mocks.checkLevelExists.mockReturnValue(false);

    const router = createMemoryRouter(routesConfig, {
      initialEntries: [`/level/${levelNum}/details`],
    });

    render(<RouterProvider router={router} />);

    expect(router.state.location.pathname).toBe("/");
  });

  it("doesn't redirect on levelNum that exists.", () => {
    const levelNum = 1;
    mocks.useParams.mockReturnValue({ levelNum });
    mocks.checkLevelExists.mockReturnValue(true);

    const router = createMemoryRouter(routesConfig, {
      initialEntries: [`/level/${levelNum}/details`],
    });

    render(<RouterProvider router={router} />);

    expect(router.state.location.pathname).toBe(`/level/${levelNum}/details`);
  });
});
