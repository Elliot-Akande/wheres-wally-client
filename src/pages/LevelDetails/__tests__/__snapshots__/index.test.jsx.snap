// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LevelDetails > displays error message on levelNum that doesn't exist. 1`] = `
<div>
  <div>
    <h1>
      Error
    </h1>
    <p>
      This level doesn't exist!
    </p>
    <a
      href="/"
    >
      Back to home page
    </a>
  </div>
</div>
`;

exports[`LevelDetails > renders title, start link, leaderboard, and home link. 1`] = `
<div>
  <div
    class="sc-beySPh dBJtdt"
  >
    <div
      class="sc-guDLey hoUZmj"
    >
      <h1
        class="sc-dmyCSP cepTqO"
      >
        LEVEL 
        3
      </h1>
      <a
        class="sc-hLQSwg kCsMut"
        href="/level/3"
      >
        Start
      </a>
    </div>
    <div>
      Leaderboard Component
    </div>
    <a
      class="sc-eDLKkx fHuTzy"
      href="/"
    >
      Back
    </a>
  </div>
</div>
`;
